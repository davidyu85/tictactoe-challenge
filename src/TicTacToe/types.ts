export interface TicTacToeProps {
  wildMode?: boolean;
  computerIsPlayer2?: boolean;
}

export type ChessPiece = 'X' | 'O';

export type TicTacToeGameState = (ChessPiece | null)[];

export type ChessAccessMap = { computer: ChessPiece[]; player: ChessPiece[] };

export type GameplayActions =
  | {
      type: 'place-chess';
      cellPos: number;
      chess: ChessPiece;
    }
  | {
      type: 'computer-place-chess';
      wildMode: boolean;
    };

export interface GameplayStates {
  gameState: TicTacToeGameState;
  isPlayerTwoTurn: boolean;
  isPlayerWins: boolean;
  isDraw: boolean;
}
