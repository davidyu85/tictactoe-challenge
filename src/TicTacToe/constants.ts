export const SINGLE_PLAYER_STRING = 'Player VS Computer';

export const TWO_PLAYER_TURN_STRING = {
  1: 'Player 1 turn',
  2: 'Player 2 turn',
};

export const SINGLE_PLAYER_WIN_STRING = {
  1: 'Player wins!',
  2: 'Computer wins!',
};

export const TWO_PLAYER_WIN_STRING = {
  1: 'Player 1 wins!',
  2: 'Player 2 wins!',
};

export const DRAW_STRING = 'Draw!';
