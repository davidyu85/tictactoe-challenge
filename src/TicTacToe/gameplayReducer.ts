import { CHESS_MAP } from './constants';
import { GameplayStates, GameplayActions, ChessPiece } from './types';
import checkEndGame from './utils/checkEndGame';
import miniMax from './utils/miniMax';

export const initState: GameplayStates = {
  gameBoard: [...Array(9).fill(null)],
  isPlayerTwoTurn: false,
  isPlayerWins: false,
  isDraw: false,
};

const gameplayReducer = (
  state: GameplayStates,
  action: GameplayActions
): GameplayStates => {
  const { gameBoard, isPlayerTwoTurn } = state;

  switch (action.type) {
    case 'place-chess': {
      const board = [...gameBoard];
      board[action.cellPos] = action.chess;

      const endGameState = checkEndGame(board, isPlayerTwoTurn);

      return {
        ...state,
        gameBoard: board,
        ...endGameState,
      };
    }

    case 'computer-place-chess': {
      const board = [...gameBoard];
      const chessMap = CHESS_MAP[action.wildMode ? 'wild' : 'standard'];
      let bestMove = { score: -Infinity, pos: -1, chess: chessMap.computer[0] };

      chessMap.computer.forEach((chess: ChessPiece) => {
        for (let pos = 0; pos < board.length; pos++) {
          if (board[pos] !== null) continue;

          board[pos] = chess;
          const score = miniMax(chessMap, board, false, -Infinity, Infinity);
          board[pos] = null;

          if (score > bestMove.score) {
            bestMove = { score, chess, pos };
          }
        }
      });

      board[bestMove.pos] = bestMove.chess;

      const endGameState = checkEndGame(board, isPlayerTwoTurn);

      return {
        ...state,
        gameBoard: board,
        ...endGameState,
      };
    }
  }
};

export default gameplayReducer;
