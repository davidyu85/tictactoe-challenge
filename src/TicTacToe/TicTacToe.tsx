import { ReactElement, useEffect, useReducer, useRef } from 'react';
import { BoardCell, GameBoard } from './Board/Board';
import gameplayReducer, { initState } from './gameplayReducer';
import { SINGLE_PLAYER_STRING, TWO_PLAYER_TURN_STRING } from './constants';
import ChessSelect from './ChessSelect/ChessSelect';
import GameEnds from './GameEnds/GameEnds';
import { ChessPiece, TicTacToeProps } from './types';

const TicTacToe = ({
  wildMode = false,
  computerIsPlayer2 = false,
}: TicTacToeProps): ReactElement => {
  const [states, dispatch] = useReducer(gameplayReducer, initState);
  const { gameBoard, isPlayerTwoTurn, isPlayerWins, isDraw } = states;
  const chessSelect = useRef<HTMLSelectElement>(null);

  const handleClickToPlaceChess = (cellPos: number) => (): void => {
    dispatch({
      type: 'place-chess',
      cellPos,
      chess: chessSelect.current?.value as ChessPiece,
    });
  };

  useEffect(() => {
    if (computerIsPlayer2 && isPlayerTwoTurn) {
      dispatch({
        type: 'computer-place-chess',
        wildMode,
      });
    }
  }, [isPlayerTwoTurn, computerIsPlayer2, wildMode]);

  const disablePlaceChess: boolean =
    isPlayerWins || isDraw || (isPlayerTwoTurn && computerIsPlayer2);

  return (
    <>
      <GameBoard>
        {gameBoard.map((chess, i) => (
          <BoardCell
            key={`cell-${i}`}
            chess={chess}
            onClickOnce={handleClickToPlaceChess(i)}
            disabled={disablePlaceChess}
          />
        ))}
      </GameBoard>

      {!computerIsPlayer2 ? (
        <label>{TWO_PLAYER_TURN_STRING[isPlayerTwoTurn ? 2 : 1]}</label>
      ) : (
        <label>{SINGLE_PLAYER_STRING}</label>
      )}

      <ChessSelect
        ref={chessSelect}
        wildMode={wildMode}
        isPlayerTwoTurn={isPlayerTwoTurn}
      />

      <GameEnds
        computerIsPlayer2={computerIsPlayer2}
        isPlayerTwoTurn={isPlayerTwoTurn}
        isPlayerWins={isPlayerWins}
        isDraw={isDraw}
      />
    </>
  );
};

export default TicTacToe;
