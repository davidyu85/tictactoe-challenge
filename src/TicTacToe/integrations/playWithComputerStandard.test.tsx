import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import TicTacToe from '../TicTacToe';
import { SINGLE_PLAYER_WIN_STRING } from '../constants';
import { renderGameStateSnapShot } from '../../../vitest.setup';

describe('Play with computer - standard mode', () => {
  beforeEach(() => {
    render(<TicTacToe computerIsPlayer2 />);
  });

  afterEach(() => {
    renderGameStateSnapShot();
  });

  it('can block a player move horizontally', () => {
    fireEvent.click(screen.getAllByRole('button')[6]);
    fireEvent.click(screen.getAllByRole('button')[7]);
    const board = screen.getAllByRole('button');

    expect(board[8].innerHTML).toStrictEqual('O');
  });

  it('can block a player move vertically', () => {
    fireEvent.click(screen.getAllByRole('button')[0]);
    fireEvent.click(screen.getAllByRole('button')[3]);
    const board = screen.getAllByRole('button');

    expect(board[6].innerHTML).toStrictEqual('O');
  });

  it('can block a player move diagonally', () => {
    fireEvent.click(screen.getAllByRole('button')[4]);
    fireEvent.click(screen.getAllByRole('button')[6]);
    const board = screen.getAllByRole('button');

    expect(board[2].innerHTML).toStrictEqual('O');
  });

  it('can play a winning move horizontally', () => {
    fireEvent.click(screen.getAllByRole('button')[3]);
    fireEvent.click(screen.getAllByRole('button')[6]);
    fireEvent.click(screen.getAllByRole('button')[7]);

    expect(screen.getByText(SINGLE_PLAYER_WIN_STRING[2])).toBeVisible();
  });

  it('can play a winning move vertically', () => {
    fireEvent.click(screen.getAllByRole('button')[1]);
    fireEvent.click(screen.getAllByRole('button')[2]);
    fireEvent.click(screen.getAllByRole('button')[5]);

    expect(screen.getByText(SINGLE_PLAYER_WIN_STRING[2])).toBeVisible();
  });

  it('can play a winning move diagonally', () => {
    fireEvent.click(screen.getAllByRole('button')[1]);
    fireEvent.click(screen.getAllByRole('button')[7]);
    fireEvent.click(screen.getAllByRole('button')[6]);

    expect(screen.getByText(SINGLE_PLAYER_WIN_STRING[2])).toBeVisible();
  });
});
