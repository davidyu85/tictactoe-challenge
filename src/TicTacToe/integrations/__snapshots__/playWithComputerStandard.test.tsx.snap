// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Play with computer - standard mode > can block a player move diagonally 1`] = `
"
 O   O
   X  
 X    
"
`;

exports[`Play with computer - standard mode > can block a player move horizontally 1`] = `
"
      
   O  
 X X O
"
`;

exports[`Play with computer - standard mode > can block a player move vertically 1`] = `
"
 X    
 X O  
 O    
"
`;

exports[`Play with computer - standard mode > can play a winning move diagonally 1`] = `
"
 O X  
   O  
 X X O
"
`;

exports[`Play with computer - standard mode > can play a winning move horizontally 1`] = `
"
 O O O
 X    
 X X  
"
`;

exports[`Play with computer - standard mode > can play a winning move vertically 1`] = `
"
 O X X
 O   X
 O    
"
`;
